### DISCUSSION TITLE
command not found

### DISCUSSION THREAD

**Discussion by zcysxy** [Taitava/obsidian-shellcommands#191] [Q&A] [answered]
Sorry I'm not that familiar with shell so I may ask some dumb questions. I want to use Picgo CLI (an image uploader) with Shell Commands, like

```shell
picgo upload
```

But it told me `zsh: command not found`. However `picgo` is in the path, and I can execute this command from the terminal.

Then I tried using the path returned by `which picgo`. But it gave me another error: `env: node: No such file or directory`.

I know this is not a bug, since another plugin Snippets Plugin gives me the same errors. But I'm confused and thought I might get some help here.

**Comment by Taitava** [comment, marked as answer]
Hi, thank you for your question! 🙂

Are you using macOS or Linux?

This is a very new problem to me, so I don't know yet what might help. There has been [another discussion with a similar issue](https://github.com/Taitava/obsidian-shellcommands/discussions/189) but that's still unsolved. I'll quote an instruction from that discussion:

> Can you try to run the following command first in the normal terminal and then using the SC plugin (you can set the output channel to be e.g. [clipboard](https://publish.obsidian.md/shellcommands/Output+handling/Output+channel+-+Clipboard)): `echo $PATH`.
> 
> If you want, you can copy the outputs into this discussion. I understand it might be private information (contains file paths), so you don't have to do it if you don't want.

You mentioned that you checked `$PATH` already, but I just wanted to make sure that did you check it by using the SC plugin?

If the content of `$PATH` in SC is different than what it is in a normal terminal, then that might give us some direction on how to find a solution to this.

Here is also another `$PATH` related discussion: https://github.com/Taitava/obsidian-shellcommands/discussions/150 . It's a feature request that is still under planning.

**Comment by zcysxy** [comment]
@Taitava Thank you so much for your reply, it solved both problems for me! I'm on macOS. And yeah `$PATH` in SC, which is `/usr/bin:/bin:/usr/sbin:/sbin`, doesn't contain the path to `picgo`. So I went to check my `.zshrc`; it turned out that I did not export the path to `picgo` to `$PATH` either. This is because I use nvm to manage node; and in `.zshrc`, a nvm script is executed, which adds the path to both `picgo` and `node` to the `$PATH`. So I just added the execution of the nvm script to the command and everything works now! The result command is this

```shell
\. "/opt/homebrew/opt/nvm/nvm.sh" && picgo upload
```

**Reply by Taitava** [reply]
Wow, nice to hear that you were able to solve the problem! 🌞 

> in `.zshrc`, a nvm script is executed, which adds the path to both `picgo` and `node` to the `$PATH`

So I assume that `.zshrc` is not executed when SC executes shell commands. SC uses Node.js's [`child_process.exec()`](https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback) to execute shell commands. So looks like the execution workflow does not include `.zshrc`, which is a pity, but maybe there is a reason for it that I'm unaware of. Maybe performance could be one reason?

I'm thinking about this from a convenience point of view: If there would be a simple, well working way to include `.zshrc` in the execution workflow, I'd like to offer users a setting to enable it. It would narrow the gap between executing shell commands via SC versus via a terminal.

**Comment by Taitava** [comment]
Just a heads up here too, #150 is now in a beta test: https://github.com/Taitava/obsidian-shellcommands/discussions/228 . It can be used to add directories to `PATH`.
